Articles/Defining and Using Observables.md

```diff
@@ -0,0 +1,346 @@
+
+## Appendix 1: Attribute Paths
+Attribute paths occur in two places: in the `observable` object's `name` attribute as a path reference to a field in the event, and in class-specific attribute observable definitions. In both cases the paths are same. (Author's note: I believe this is the only use of JSON path-like capability in OCSF.)
+
+The general pattern is dot-separated attribute names, for example `foo.bar`. Using the dot (".") as a separator works well because OCSF does not use dots in attribute names. There is no special notation for arrays, so these paths only tell us that a reference is for _one of_ the items along a path that includes one or more arrays.
+
+These attribute path references are similar (at least in spirit) to [JSON Pointer](https://www.rfc-editor.org/rfc/rfc6901), [JSONPath](https://www.rfc-editor.org/rfc/rfc9535), and the syntax used by the [`jq` command-line tool](https://github.com/jqlang/jq), though simpler and notably without array notation.
+
+Let's say we have a event with a nested structure as follows:
+```jsonc
+{
+  // ... other event fields
+  "devices": [
+    {
+      "hostname": "mercury",
+      "network_interfaces": [
+        {
+          "ip": "10.0.0.5"
+        }
+      ]
+    },
+    {
+      "hostname": "venus",
+      "network_interfaces": [
+        {
+          "ip": "192.168.0.3"
+        },
+        {
+          "ip": "10.100.0.42"
+        }
+      ]
+    }
+  ]
+}
+```
+
+In this example, `ip` is an observable by dictionary type with `type_id` of `2`. The following shows what the observables for this event look like:
+
+```jsonc
+{
+  // ... other event fields
+  "devices": [
+    // ... as above
+  ],
+  "observables": [
+    {
+      "name": "devices.network_interfaces.ip",
+      "type_id": 2,
+      "value": "10.0.0.5"
+    },
+    {
+      "name": "devices.network_interfaces.ip",
+      "type_id": 2,
+      "value": "192.168.0.3"
+    },
+    {
+      "name": "devices.network_interfaces.ip",
+      "type_id": 2,
+      "value": "10.100.0.42"
+    }
+  ]
+}
+```
+
+Notice that the attribute path in each `name` field is the same; the positions in the `devices` and `network_interfaces` arrays is not included.
+
+> [!IMPORTANT]
+> **TODO for reviewers:**
+>
+> Should we include array syntax? We arrays could be added to the existing dot syntax. E.g., `foo.bars[0].baz.quxes[3].value` instead of `foo.bars.baz.quxes.value`. Alternately, we could adopt JSON Pointer syntax.
```

Comment:
I am specifically interested in the outcome of this TODO/uncertainty. 

I like the approach proposed by the document (e.g. `devices.network_interfaces.ip`). For me the reason for `name` field is to provide additional information about the meaning of observable. That is sufficiently done by that syntax.

The option with specific indices (e.g. `devices[0].network_interfaces[1].ip`) seem redundant. The indices are not important for the meaning (e.g. first device has no different meaning than second). If someone would like to use the name for indexing/searching, then it's straightforward to match `name = "devices.network_interfaces.ip"`, but it would be tricky with the indices.

The in-between option with just brackets (e.g. `devices[].network_interfaces[].ip`) gives additional type information, but it's not clear what is it's value. With having the schema at hand, this information could be obtained from schema. Even without the schema, it's possible to correctly traverse (or search) the data without the additional `[]` type hint. When entered manually it might be error prone - it's easy to forget '[]'.

Reply:
I changed this from a TODO in the doc to a review comment. I'm leaning toward not introducing an array index syntax here as well.